relays/ethereum/src/ethereum_client.rs

```diff
@@ -150,6 +214,172 @@ async fn transaction_receipt(client: Client, hash: H256) -> (Client, Result<Rece
+			data: Some(encoded_call.into()),
+		})
+		.map_err(|e| Error::RequestSerialization(e)),
+		client
+	);
+	let (client, call_result) =
+		bail_on_error!(call_rpc::<Bytes>(client, "eth_call", Params::Array(vec![call_request]),).await);
+	match call_decoder.decode(&call_result.0) {
+		Ok(is_known_block) => (client, Ok((id, is_known_block))),
+		Err(error) => (client, Err(Error::ResponseParseFailed(format!("{}", error)))),
+	}
+}
+
+/// Submits Substrate headers to Ethereum contract.
+pub async fn submit_substrate_headers(
+	client: Client,
+	params: EthereumSigningParams,
+	contract_address: Address,
+	headers: Vec<QueuedSubstrateHeader>,
+) -> (Client, Result<Vec<SubstrateHeaderId>, Error>) {
+	let (mut client, mut nonce) =
+		bail_on_error!(account_nonce(client, params.signer.address().as_fixed_bytes().into()).await);
+
+	let ids = headers.iter().map(|header| header.id()).collect();
+	for header in headers {
+		let (ret_client, _) = bail_on_error!(
+			submit_ethereum_transaction(
+				client,
+				&params,
+				Some(contract_address),
+				Some(nonce),
+				false,
+				bridge_contract::functions::import_header::encode_input(header.extract().0.encode(),),
+			)
+			.await
+		);
+
+		nonce += 1.into();
+		client = ret_client;
+	}
+
+	(client, Ok(ids))
+}
+
+/// Deploy bridge contract.
+pub async fn deploy_bridge_contract(
+	client: Client,
+	params: &EthereumSigningParams,
+	contract_code: Vec<u8>,
+	initial_header: Vec<u8>,
+	initial_set_id: u64,
+	initial_authorities: Vec<u8>,
+) -> (Client, Result<(), Error>) {
+	submit_ethereum_transaction(
+		client,
+		params,
+		None,
+		None,
+		false,
+		bridge_contract::constructor(contract_code, initial_header, initial_set_id, initial_authorities),
+	)
+	.await
+}
+
+/// Submit ethereum transaction.
+async fn submit_ethereum_transaction(
+	client: Client,
+	params: &EthereumSigningParams,
+	contract_address: Option<Address>,
+	nonce: Option<U256>,
+	double_gas: bool,
+	encoded_call: Vec<u8>,
+) -> (Client, Result<(), Error>) {
+	let (client, nonce) = match nonce {
+		Some(nonce) => (client, nonce),
+		None => bail_on_error!(account_nonce(client, params.signer.address().as_fixed_bytes().into()).await),
+	};
+	let (client, gas) = bail_on_error!(
+		estimate_gas(
+			client,
+			CallRequest {
+				to: contract_address,
+				data: Some(encoded_call.clone().into()),
+			}
+		)
+		.await
+	);
+	let raw_transaction = ethereum_tx_sign::RawTransaction {
+		nonce,
+		to: contract_address,
+		value: U256::zero(),
+		gas: if double_gas { gas.saturating_mul(2.into()) } else { gas },
+		gas_price: params.gas_price,
+		data: encoded_call,
+	}
+	.sign(&params.signer.secret().as_fixed_bytes().into(), &params.chain_id);
+	let transaction = bail_on_arg_error!(
+		to_value(Bytes(raw_transaction)).map_err(|e| Error::RequestSerialization(e)),
+		client
+	);
+	let (client, _) = bail_on_error!(
+		call_rpc::<TransactionHash>(client, "eth_submitTransaction", Params::Array(vec![transaction])).await
```

Comment:
To be honest I'd really prefer to enclose low-level non-bridge-sepcific (so no `EthereumSigningParams`) calls into a separate module and have it properly typed:
```rust
mod ethereum_rpc {
  pub async fn submitTransaction(client: _, transaction: RawTransaction, secret: _) -> (Client, Result<_>) {
   ...
  }
  pub async fn call(client: _, CallRequest) -> .. { }
}
```
Right now it feels quite error prone to duplicate the same logic in multiple places (especially eth_call) and use stringified methods.

The higher-level (bridge-specific), like: `deploy_bridge`, `best_substrate_block` should be exposed from this module, but I'd keep the low-level ones encapsulated, so that it's easier to distinguish bridge-specific logic from RPC call boilerplate (like encoding the `CallRequest`).

Reply:
Logged here: https://github.com/paritytech/parity-bridges-common/issues/74